Feature/add storage

MoonSeongKwang opened the pull request:
Chain Plugin에 Strorage 관련 파일들을 merge합니다.

큰 변경점은 다음과 같습니다.
1. Storage 관련 파일들이 추가됩니다. 현재 구동은 블럭 정보를 담고 있는 json 파일을 파싱하여 처리가 되는것까지 확인하였으며, 저장까지는 아직 테스트하지 않았습니다. 이 커밋 이후 바로 테스트하여 추가할 예정입니다.

2. Key-Value Store를 담당하는 levelDB가 lib/ 에 추가됩니다.
변경된 파일을 확인하실 때, levelDB 관련 파일들은 전부 무시하셔도 됩니다.

그 외 변경사항 등은 각 커밋에 정리된 것이 있으니, 참조하시고 명확하지 않은 것이 있다면 추가로 질문해주세요.

MoonSeongKwang committed:
# ADD - Storage file add / edit storage_type
우선은 구조를 생각하지 않고 include 경로만 맞춰서 hpp는 include에,
cpp는 본 디렉토리에 위치하여 파일을 삽입하였습니다.
storage_type는 다른 작업을 전혀 하지 않는 파일이므로 config에
포함시켰으며, 여기로 migrate하면서 enterprise에 있었고 당장
storage에서는 쓰이지 않던 옵션들을 삭제하였습니다.

추후 필요한 구조에 따라 파일들의 경로가 바뀔 예정입니다.

MoonSeongKwang committed:
# EDIT - Directory structure change and etc.
변경 - 파일들의 위치를 바꾸었습니다.
structure/ 디렉토리에는 block, transaction, cert, endorser, sigature 과
같은 특정 구조를 나타내는 파일들이 위치합니다. 이들은 특정 알고리즘은
거의 포함하고있지 않으며, 해당 구조에 대한 연산을 주로 수행합니다.

MoonSeongKwang committed:
# ADD - DB controller add
하위 함수들에서 DB에 접근해야 하는데, Chain.hpp/cpp의 private에 세션이
유지되면 접근하기가 어렵습니다. 그래서 구조를 변경합니다.

 - Chain.cpp/hpp의 db_session 변수 삭제. 그로 인한 생성자 수정.
 - chain_plugin/include/db_controller 추가.
DBController 클래스의 public 변수로 m_db_pool을 유지합니다.
이곳에서 db connection을 pool로 유지하고 있으므로, 아래의 사용법에 따라 사용하시면 됩니다.

 ## Usage
1. controller 선언
DBController db_controller("service", "dbname", "username", "password");
2. session pool에서 가져옴
soci::session db_session(db_controller.m_db_pool);
3. 쿼리 실행
soci::statement st = (db_session.prepare << "something query", soci::into(soci::row result));
st.execute(true);

MoonSeongKwang committed:
# ADD - Plugin setting, EDIT - DBController, etc.
1. Plugin 구조에 맞추기 시작했습니다.
아직은 지속적으로 메시지가 들어오는것을 감시하는 등의 구현은 되어있지
않아서 단순히 존재하는 파일에서 읽어오는 것을 확인한 단계이지만,
알고리즘 로직의 구동 확인은 가능합니다.

2. DB 컨트롤러를 전반적으로 재설계했습니다. 기존의 db_lib에서 db 접근을
컨트롤했는데, 세션을 플러그인 쪽에서 관리해야하기 때문에 db_lib의 쿼리
실행과 같은 기능을 모두 db_controller로 통일했습니다.
커넥션은 pool 형태로 관리되며, config 파일에 풀의 크기 설정이 있습니다.

3. genesis-block 파일을 world-create로 변경하였습니다. 내부에 있는
정보들이 블록 자체에 대한 정보라기보다는 월드와 체인의 생성에 관한
내용이기 때문입니다.

4. first_block_test.json 파일을 추가하였습니다.
아직은 직접적으로 메시지 통신에 의한 데이터 처리를 테스트할 단계는
아니기에, 메시지의 형태를 가진 파일을 읽어오는 것으로 처리를 대체합니다.
트랜잭션의 txagg_cbor 관련 정보는 정확히 생성하지 않아서, 관련 정보를
처리하는 코드 또한 임시로 주석처리 되어있습니다. 다음 테스트때에는
트랜잭션에 대한 정보까지 정확하게 수정할 예정입니다.

5. block에서 user certification을 처리하는 부분을 수정하였습니다.
pem이 여러 개의 string 형태의 배열로 오기 때문에, 관련 처리를 해야
합니다.
pk 형식은 한 줄로 된 정보일거라고 하시기 때문에, 추후 pk/pem이 둘 다 올
수 있는 항목들에 대해서는 처리가 필요해 보입니다.

6. 로그가 기존의 easylog로 남아있는 것들이 많아서 public의 spdlog쪽의
함수로 변경했습니다. 단, 아직 완전한 정상 작동은 보장하지 않습니다.

7. levelDB를 추가하였습니다. 아직 Key-Value Store쪽은 구현되지
않았습니다.

blackironj commented:
`leveldb`  lib에 포함 하지말고, 설치해서 쓰는게 어떨까요? 혹시 GEN 할때 문제있었나요?

SeongKwangMoon commented:
저도 lib에 포함할지 고민하다가 여쭤보고 일단 포함시킨 거라서, 논의 후에 바꾸면 될 것 같습니다.
conflict는 지금 확인중입니다.

MoonSeongKwang committed:
Merge branch 'develop' into feature/add-storage
 # FIX - conflict fix

blackironj commented on src/plugins/chain_plugin/config/config.hpp:
```diff
@@ -0,0 +1,50 @@
+
+// SETTING
+
+        constexpr size_t MAX_SIGNER_NUM = 200;
```
되도록 size_t 안쓰는게 좋을거 같아요.

blackironj commented on src/plugins/chain_plugin/config/storage_type.hpp:
```diff
@@ -0,0 +1,41 @@
+constexpr auto TRANSACTION_ID_TYPE_SIZE = 32;
+constexpr auto ALPHANUMERIC_ID_TYPE_SIZE = 8;
+
+using block_height_type = size_t;
```
여기도 size_t 요

park53kr commented:
번거롭지 않다면 저는 설치해서 사용해도 상관없습니다.

MoonSeongKwang committed:
# FIX - size_t to int or int32_t
size_t가 사용된 부분들을 확인하여 다른 자료형으로 교체하였습니다.

blackironj commented:
@park53kr @SeongKwangMoon 
패키지 매니저에서 leveldb 1.20은 설치가 가능한거 같네요. 
cmake에서 find_package가 지원이 안돼서 따로 작성해야하는거 할 필요가 있을거같아요. 지금 어떻게 하는게 좋을까요?
`apt install libleveldb-dev`
`brew install leveldb`

park53kr commented:
그럼 find_package 관련 코드 작성하면 되지 않을까요?

blackironj commented:
https://github.com/BVLC/caffe/blob/master/cmake/Modules/FindLevelDB.cmake
이런거 고쳐쓰면 될 거 같습니다.

SeongKwangMoon commented:
그러면 일단 lib/ 쪽에서는 빼놓으면 될까요? 나중에 추가하기로 해도 그 때 다시 추가하면 될테니까요

blackironj commented:
@SeongKwangMoon 
현재 leveldb 관련 파일 lib에서 빼고, `FindLevelDb.cmake` 작성하고 leveldb는 설치해서 쓰는게 좋을거 같아요. 패키지 매니저로 설치할 수 있으니까.

MoonSeongKwang committed:
# DELETE - lib/leveldb
package로 설치가 가능하므로 프로젝트 내에서 포함시키는것은 제외하고,
FindLevelDB.cmake로 찾아서 빌드할 수 있도록 수정하였습니다.
추후 leveldb 관련 구현을 할 때 FindLevelDB.cmake가 정상 작동하는지를 다시 한 번
확인할 예정입니다.

MoonSeongKwang committed:
# EDIT - first_block_test.json fix, etc.
first_block_test.json의 tx_cbor 관련 내용을 유효한 값으로 설정한 후 출력
테스트까지 완료하였습니다.
또, min 함수에서 type 관련 빌드 에러가 생겨서 int32_t를 uint32_t로 변경하였습니다.

blackironj reviewed (approved):
`FindLevedlDB.cmake` 는 수정이 필요해 보여요. 말씀하신대로 차후 leveldb 관련 작업하시면서 수정해주시면 될거 같습니다.

MoonSeongKwang committed:
# FIX - minor fix
cmake에 lib/leveldb 경로가 남아있는것을 삭제하였고,
storage_config.hpp에 GENESIS_BLOCK_PREV_ID가 B64 형태로 되어있던것을 B58
형태로 수정하였습니다.

morishjs commented on src/plugins/chain_plugin/CMakeLists.txt:
```diff
@@ -1,25 +1,44 @@
+
+file(GLOB SOURCES
+        *.cpp
+        include/*.cpp
```
필요없는 코드입니다.
include/ 디렉토리에는 hpp만 들어가요

morishjs commented on src/plugins/chain_plugin/include/static_merkle_tree.hpp:
```diff
@@ -0,0 +1,152 @@
+  void generate(vector<hash_t> &merkle_contents) {
+    const bytes dummy_leaf(32, 0); // for SHA-256
+
+    auto min_addable_size = min(static_cast<uint64_t>(MAX_MERKLE_LEAVES), merkle_contents.size());
```
Merkle_content.size() 가 size_t 이므로, uint64_t로 형변환해야 합니다.

morishjs reviewed (changes requested):
실제로 빌드해서 돌려보면 런타임에러가 발생합니다.

MoonSeongKwang committed:
# FIX - Cmake, size_t func fix
cmake의 불필요했던 include/*.cpp와
size_t 형태로 return하는 결과값의 형변환을 실시했습니다.

SeongKwangMoon commented:
@park53kr 
 런타임 에러가 발생하나요? 저는 지금 Chain만 활성화해서 실행 테스트해보고있는 중이긴 한데, Chain 플러그인만 활성화했을 때에는 런타임 에러가 발생하지 않는데 혹시 다른 부분과의 충돌이 있는걸까요?

park53kr commented:
@SeongKwangMoon 
아 그럼 다른 플러그인에서 에러가 발생하는 것 같습니다. 이 PR에서는 신경을 안 쓰셔도 될 것 같습니다.
